Fix window positioning on Windows when the taskbar is on the top or left

logiclrd opened the pull request:
In Windows versions prior to Windows 11, the Task Bar can be docked to any edge of the screen. Even with Windows 11, this can still be done with third-party tools such as "StartAllBack". 

When RustDesk creates its main window, it (currently) supplies a hardcoded window position of (10, 10):

https://github.com/rustdesk/rustdesk/blob/317639169359936f7f9f85ef445ec9774218772d/flutter/windows/runner/main.cpp#L129

If the Task Bar is on the top or left edge of the screen, the pixel offset (10, 10) might be covered by the task bar. Part of the RustDesk window is thus not visible and inaccessible:

<img width="1275" height="477" alt="image" src="https://github.com/user-attachments/assets/b55745f8-7a81-4ee7-8754-8fb4a151c9e1" />

With no visible title bar, the window cannot be dragged to reposition it, and window control widgets located in the title bar are inaccessible. (There is a non-obvious workaround; when a window resize operation completes, Windows automatically crops the resulting rectangle to the active work area, so using any of the visible edges to resize the window causes the top edge of the window to be moved down making the title bar visible.)

This PR addresses this by querying the _work area_. This indicates the exact coordinate range that is _not_ covered by permanent fixtures such as the Task Bar. The `origin` and `size` are then computed based on this.

The work area is in _virtual_ pixels. If the system desktop scale is not 100%, then the work area is correspondingly scaled. For instance, I'm presently using a 50" TV as my display, native resolution 1920x1080, and if I run it at the standard assumed 96 dpi, the text is difficult to read. So, my desktop scale ("Make everything bigger" in Display settings) is set to 150%. As a result, the work area is returned as (0, 40)-(1280, 720), instead of the actual literal pixel area of (0, 60)-(1920, 1080).

But, this is okay, because the `Win32Window` abstraction being used has, in its `CreateAndShow` method, code to find the monitor's DPI and scale the coordinates correspondingly. So, the coordinate system for `origin` is the same as the scaled frame of reference used for the work area (*as long as `origin` is on the primary monitor -- if not, they might be the same but they might not, but in any case the current hardcoded (10, 10) will always be on the primary monitor).

I wasn't able to get a build environment set up. The `rustup_init.exe` tool created a `bin` folder with all the requisite `.exe` files, except they're all 0-byte files. Not sure why. So, I tested this code piecemeal, by copy/pasting it into a smaller project that I can build. It worked there. But, it might require tweaks in situ. I know that's a pain in the butt, and if I figure out why Rust didn't install properly, I'll see if I can't build it locally. In the meantime, I'm hoping someone with a working build environment will take pity on me and do a local build and push any fixes. :-)

logiclrd committed:
Added win32_desktop.cpp/.h defining a method Win32Desktop::GetWorkArea.
Added code to wWinMain in main.cpp to position the window relative to the work area, which may not be at (0, 0) depending on the user's configuration.

rustdesk commented:
@fufesou

fufesou commented:
Maybe a better way would be to center the main window on the first run?

logiclrd commented on flutter/windows/runner/win32_desktop.cpp:
```diff
@@ -0,0 +1,27 @@
+  {
+    RECT workAreaRect;
+
+    if (!SystemParametersInfoA(SPI_GETWORKAREA, 0, &workAreaRect, 0))
```
There is no function `SystemParametersInfo`. It's just a macro that calls either `SystemParametersInfoA` or `SystemParametersInfoW`. The parameter in this instance doesn't involve any string values at all, so it is irrelevant which entrypoint is used. I noticed that other existing code uses the ANSI functions for Win32 API calls, such as:

* `LoadLibraryA` in `main.cpp`, `win32_window.cpp`
* `GetEnvironmentVariableA` in `wf_cliprdr.c`
* `LoadLibraryExA`, `GetModuleHandleExA`, `LoadCursorA`, `RegisterClassExA`, `CreateWindowExA`, `FindWindowExA` in  `mag.rs`
* `FormatMessageA` in `lddController.c`
* `ShellExecuteA` in `virtual_display_manager.rs`

These existing ANSI calls were what I immediately saw in the surrounding code, so I used `SystemParametersInfoA` to follow suit.

But, I do recognize that using the wide-character functions is generally preferable. It avoids an extra layer of indirection in the use of the thunk, and if actual character strings are involved, it also avoids issues where the ANSI code page might not support characters being used, making it impossible to access certain files, properly display messages, etc. *Those concerns are irrelevant to a call to `SystemParametersInfo` with action `SPI_GETWORKAREA`.*

Should I change this call to use the wide-character entrypoint, or leave it using the ANSI endpoint for consistency with adjacent code?

logiclrd committed:
Corrected the constraint on the size value calculated by main.cpp.

fufesou commented:
> Property centering the window still requires calculations based on the work area. It's not a bad idea, but it does feel out-of-scope for this PR.

Yes. But maybe there is no need to calculate the work area.

RustDesk will call `restoreWindowPosition()` after the window is initialized.

If the patch is applied https://github.com/rustdesk/rustdesk/pull/12933#issuecomment-3299211448 , then no need to do the calculation in  `rustdesk/flutter/windows/runner/main.cpp`.

logiclrd commented:
From where I'm sitting, calling a 20-line function that calls a system API that only deals in numbers and (probably) can't fail "overkill" is a bit weird, but okay. :-) I am already pursuing correcting `WindowManager::center` upstream. At present, they aren't centered correctly, but that isn't the fault of this codebase.

fufesou commented:
Win7, Win10, Win11 are tested, the work areas are correct.

Please:
1. Fix the build issue, `min()` should be `std::min()`.
2. Maybe centering the main window is also needed. https://github.com/rustdesk/rustdesk/pull/12933#issuecomment-3299211448

logiclrd committed:
Fixed references to min to use std::min.

logiclrd committed:
Reworked GetWorkArea in win32_desktop.cpp to treat the supplied origin and size as containing an existing window rectangle, and to find the monitor that contains or is closest to that window.
Added function FitToWorkArea to win32_desktop.cpp/.h.
Updated main.cpp to use Win32Desktop::FitToWorkArea instead of explicitly constraining the size.

logiclrd commented:
In consideration of the distinct possibility of adding code to persist the window bounds in the future, I've made some architectural changes. It's still within the bounds of scope for this PR, but there is now a function `Win32Desktop::FitToWorkArea` that encapsulates exactly the work needed to take a previously-saved set of bounds and make sure they're still sane.

I will shortly rebase the `centre-main-window` branch ontop of this.
